src/ip.cpp

```diff
@@ -4,6 +4,21 @@
 
 namespace oxen::quic
 {
+    ipv4::ipv4(const std::string& str)
+    {
+        struct sockaddr_in sin
+        {};
+        sin.sin_family = AF_INET;
+        auto rv = inet_pton(AF_INET, str.c_str(), &sin.sin_addr);
+
+        if (rv == 0)  // inet_pton returns this on invalid input
+            throw std::invalid_argument{"Cannot construct address: invalid IP"};
+        if (rv < 0)
+            throw std::system_error{errno, std::system_category()};
```

Comment:
This is duplicating some code (mostly the error-to-exception translation) from Address (and also duplicating some code across the v4/v6 versions) that would be nicer to DRY out.  I suggest we extract this into something like this:

```c++
// Wrapper around inet_pton that throws an exception on error
void parse_addr(int af, void* dest, const std::string& from) {
    auto rv = inet_pton(af, from.c_str(), dest);
    // ... check rv and throw if bad
}
// Parses an IPv4 address from string
void parse_addr(in_addr& into, const std::string& from) {
    parse_addr(AF_INET, &into.sin_addr, from);
}
// Parses an IPv6 address from string
void parse_addr(in6_addr& into, const std::string& from) {
    parse_addr(AF_INET6, &into.sin6_addr);
}
```
and then this code becomes just
```c++
    in_addr sin_addr;
    parse_addr(sin_addr, str);
    addr = oxenc::big_to_host(sin_addr.s_addr);
```
and similarly for the `ipv6` and `Address` string constructors.

Reply:
great suggestion